ignite/metrics/vision/object_detection_average_precision_recall.py

```diff
@@ -0,0 +1,482 @@
+    _scores: List[torch.Tensor]
+    _y_pred_labels: List[torch.Tensor]
+    _y_true_count: torch.Tensor
+    _num_classes: int
+
+    def __init__(
+        self,
+        iou_thresholds: Optional[Union[Sequence[float], torch.Tensor]] = None,
+        rec_thresholds: Optional[Union[Sequence[float], torch.Tensor]] = None,
+        num_classes: int = 91,
+        max_detections_per_image_per_class: int = 100,
+        area_range: Literal["small", "medium", "large", "all"] = "all",
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+    ) -> None:
+        r"""Calculate mean average precision & recall for evaluating an object detector in the COCO way.
+
+
+        Average precision is computed by averaging precision over increasing levels of recall thresholds.
+        In COCO, the maximum precision across thresholds greater or equal a recall threshold is
+        considered as the average summation operand; In other words, the precision peek across lower or equal
+        sensivity levels is used for a recall threshold:
+
+        .. math::
+            \text{Average Precision} = \sum_{k=1}^{\#rec\_thresholds} (r_k - r_{k-1}) max(P_{k:})
+
+        Average recall is the detector's maximum recall, considering all matched detections as TP,
+        averaged over classes.
+
+        Args:
+            iou_thresholds: sequence of IoU thresholds to be considered for computing mean average precision & recall.
+                Values should be between 0 and 1. If not given, COCO's default (.5, .55, ..., .95) would be used.
+            rec_thresholds: sequence of recall thresholds to be considered for computing mean average precision.
+                Values should be between 0 and 1. If not given, COCO's default (.0, .01, .02, ..., 1.) would be used.
+            num_classes: number of categories. Default is 91, that of the COCO.
+            area_range: area range which only objects therein are considered in evaluation. By default, 'all'.
+            max_detections_per_image_per_class: maximum number of detections per class in each image to consider
+                for evaluation. The most confident ones are selected.
+            output_transform: a callable that is used to transform the :class:`~ignite.engine.engine.Engine`'s
+                ``process_function``'s output into the form expected by the metric. An already provided example is
+                :func:`~ignite.metrics.vision.object_detection_average_precision_recall.coco_tensor_list_to_dict_list`
+                which accepts `y_pred` and `y` as lists of tensors and transforms them to the expected format.
+                Default is the identity function.
+            device: specifies which device updates are accumulated on. Setting the
+                metric's device to be the same as your ``update`` arguments ensures the ``update`` method is
+                non-blocking. By default, CPU.
+        """
```

Comment:
Let's add [.. versionadded::](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded) tags to all public classes